File: quizzes3/04-problem.ts
export {};

type Customer = {
    name: string;
    email: string;
};

type CustomerMap = {
    data: Customer;
    addedToDup: boolean;
};

const customers: Customer[] = [
    { name: 'john doe', email: 'john@example.com' },
    { name: 'John Doe', email: 'JOHN@example.com' },
    { name: 'jane smith', email: 'jane@example.com' },
];

const findDuplicatesWithLoop = (): Customer[] => {
    const seen = new Map<string, CustomerMap>();
    const duplicateCustomers: Customer[] = [];

    for (const customer of customers) {
        const key: string = `${customer.name.toLowerCase()}|${customer.email.toLowerCase()}`;
        const entry = seen.get(key);

        if (!entry) {
            seen.set(key, { data: customer, addedToDup: false });
            continue;
        }

        if (!entry.addedToDup) {
            duplicateCustomers.push(entry.data);
            entry.addedToDup = true;
        }

        duplicateCustomers.push(customer);
    }

    return duplicateCustomers;
};

console.log(JSON.stringify({ data: findDuplicatesWithLoop() }, null, 2));

File: bins2/04-problem.rb
# frozen_string_literal: true

require "json"

customers = [
  { "name": "john doe", "email": "john@example.com" },
  { "name": "John Doe", "email": "JOHN@example.com" },
  { "name": "jane smith", "email": "jane@example.com" }
]

def self.find_duplicates_with_loop(customers)
  seen = {}
  duplicate_customers = []

  customers.each do |customer|
    key = "#{customer[:name].downcase}|#{customer[:email].downcase}"
    entry = seen[key]

    unless entry
      seen[key] = { data: customer, added: false }
      next
    end

    unless entry[:added]
      duplicate_customers << entry[:data]
      entry[:added] = true
    end

    duplicate_customers << customer
  end

  duplicate_customers
end

puts JSON.pretty_generate({ data: find_duplicates_with_loop(customers) })
